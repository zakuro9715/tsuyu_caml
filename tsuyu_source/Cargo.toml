[package]
name = "tsuyu_source"
version = "0.1.0"
edition = "2021"

[dependencies]
nanoid = "0.4.0"
tsuyu_utils = { path = "../tsuyu_utils" }
tsuyu_chars = { path = "../tsuyu_chars" }

[dev-dependencies]
tsuyu_asserts = { path = "../tsuyu_asserts" }
