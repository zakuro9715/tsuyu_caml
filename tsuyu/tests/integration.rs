use insta::{assert_toml_snapshot, glob};
use tsuyu_source::Source;
use serde::Serialize;
use std::{path::Path, process};

mod tsuyu_testutil {
    use super::*;
    pub fn walk_files<F: FnMut(&str, &Path)>(dir: &str, ext: &str, mut f: F) {
        glob!(&format!("{}/*{}", dir, ext), |path| insta::with_settings!({
            snapshot_path => dir,
            prepend_module_to_snapshot => false,
            snapshot_suffix => "",
            input_file => path.to_path_buf(),
        }, {
            let name = path.file_name().unwrap().to_string_lossy();
            f(&name, path);
        }));
    }

    pub fn run_source(path: impl AsRef<Path>) -> Output {
        let source = Source::read_file(path.as_ref()).unwrap();
        Output::from(tsuyu::run(source).unwrap())
    }
}

#[test]
fn correct_sources() {
    let sources = "sources";
    tsuyu_testutil::walk_files(sources, ".c", |name, path| {
        assert_toml_snapshot!(name, tsuyu_testutil::run_source(path));
    });
}

#[test]
fn compile_error() {
    tsuyu::compile(Source::inline("xx")).expect_err("");
}

#[derive(Serialize)]
pub struct Output {
    pub stdout_text: String,
    pub stdout_bytes: Vec<u8>,
    pub stderr_text: String,
    pub stderr_bytes: Vec<u8>,
    pub code: Option<i32>,
}

impl From<process::Output> for Output {
    fn from(output: process::Output) -> Self {
        fn parse_out(data: Vec<u8>) -> (String, Vec<u8>) {
            match String::from_utf8(data) {
                Ok(text) => (text, vec![]),
                Err(err) => (String::new(), err.into_bytes()),
            }
        }

        let (stdout_text, stdout_bytes) = parse_out(output.stdout);
        let (stderr_text, stderr_bytes) = parse_out(output.stderr);
        Self {
            stdout_text,
            stdout_bytes,
            stderr_text,
            stderr_bytes,
            code: output.status.code(),
        }
    }
}
