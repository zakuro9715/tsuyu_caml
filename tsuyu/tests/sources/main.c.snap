---
source: tsuyu/tests/integration.rs
expression: "tsuyu_testutil::run_source(path)"
input_file: tsuyu/tests/sources/main.c
---
stdout_text = '''
10
'''
stdout_bytes = []
stderr_text = ''
stderr_bytes = []
code = 0
